service/create/loadbalancer.go

```diff
@@ -0,0 +1,172 @@
+	}
+
+	// create DNS record for LB
+	hzName, err := hostedZoneName(cluster)
+	if err != nil {
+		return microerror.MaskAny(fmt.Errorf("could not generate hosted zone name: %s", err))
+	}
+
+	hz := awsresources.HostedZone{
+		Name:   hzName,
+		Client: clients.Route53,
+	}
+
+	hzCreated, err := hz.CreateIfNotExists()
+	if err != nil {
+		return microerror.MaskAny(fmt.Errorf("error creating hosted zone '%s'", errgo.Details(err)))
+	}
+
+	if hzCreated {
+		s.logger.Log("debug", fmt.Sprintf("created hosted zone"))
+	} else {
+		s.logger.Log("debug", fmt.Sprintf("hosted zone already existed, reusing"))
+	}
+
+	recordSet := &awsresources.RecordSet{
+		Client:       clients.Route53,
+		Resource:     lb,
+		Domain:       cluster.Spec.Cluster.Kubernetes.API.Domain,
+		HostedZoneID: hz.ID(),
+	}
+
+	if err := recordSet.CreateOrFail(); err != nil {
+		return microerror.MaskAny(fmt.Errorf("error registering DNS '%s'", errgo.Details(err)))
+	}
+
+	s.logger.Log("debug", fmt.Sprintf("created or reused DNS record for ELB"))
+
+	s.logger.Log("debug", fmt.Sprintf("waiting for masters to be ready..."))
+	var awsFlavouredMasterIDs []*string
+
+	for _, masterID := range masterIDs {
+		awsFlavouredMasterIDs = append(awsFlavouredMasterIDs, aws.String(masterID))
+	}
+
+	if err := clients.EC2.WaitUntilInstanceRunning(&ec2.DescribeInstancesInput{
+		InstanceIds: awsFlavouredMasterIDs,
+	}); err != nil {
+		return microerror.MaskAny(fmt.Errorf("masters took too long to get running, aborting: %v", err))
+	}
+
+	if err := lb.RegisterInstances(masterIDs); err != nil {
+		return microerror.MaskAny(fmt.Errorf("could not register instances with LB: %s", errgo.Details(err)))
+	}
+
+	s.logger.Log("debug", fmt.Sprintf("instances registered with ELB"))
+
+	return nil
+}
+
+func (s *Service) deleteLoadBalancer(clients awsutil.Clients, cluster awstpr.CustomObject) error {
+	// Delete ELB
+	lb, err := awsresources.NewExistingELB(cluster.Spec.Cluster.Cluster.ID, clients.ELB)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	if err := lb.Delete(); err != nil {
+		return microerror.MaskAny(err)
+	}
+	s.logger.Log("debug", "deleted ELB")
+
+	hzName, err := hostedZoneName(cluster)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	hz, err := awsresources.NewExistingHostedZone(hzName, clients.Route53)
+	if err != nil {
+		switch err.(type) {
+		case awsresources.DomainNamedResourceNotFoundError:
+			s.logger.Log("warn", "could not find hosted zone to delete, continuing")
+		default:
+			return microerror.MaskAny(err)
+		}
+	}
+
+	// Delete DNS record
+	recordSet := &awsresources.RecordSet{
+		Client:       clients.Route53,
+		Resource:     lb,
+		Domain:       cluster.Spec.Cluster.Kubernetes.API.Domain,
+		HostedZoneID: hz.ID(),
+	}
+
+	if err := recordSet.Delete(); err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	s.logger.Log("debug", fmt.Sprintf("deleted DNS entry for ELB"))
+
+	// Delete HZ
+	if err := hz.Delete(); err != nil {
```

Comment:
This feels weird to me. Are we going to have to redelegate the HZ whenever we create a cluster? I would expect the HZ to be kept around.